docs/odcs.rst

```diff
@@ -0,0 +1,123 @@
+rebuilds pick up the latest, most relevant, signed packages.
+
+The On Demand Compose Service (`odcs`_) project has been created to address
+current limitations in obtaining a yum repo with signed packages. The general
+idea is to not require the user to provide a yum repo url when building
+container images.  Instead, OSBS will use ODCS to generate a yum repo with the
+relevant signed packages based on user configuration.
+
+API Changes
+-----------
+
+The following parameters will be used by ``buildContainer`` task defined by
+``koji-containerbuild`` plugin:
+
+``odcs_signed <bool, default True>`` When set to False, OSBS will provide an
+empty list of ``sigkeys`` to ODCS. An empty list of ``sigkeys`` indicates the
+signature of packages will not be checked causing the compose to potentially
+include unsigned packages. If ODCS integration is not in place, this parameter
+has no effect.
+
+``container.yaml`` Changes
+--------------------------
+
+In git repo, alongside Dockerfile, the ``container.yaml`` file will be enhanced
+to contain a "compose" section. This section will include all the required
+information for requesting a compose to be created by ODCS::
+
+    compose:
+      # Optional, default False. Inherit compose used for building parent image
+      inherit: true
+      # Optional, default "tag".
+      type: tag
+      # Required for "tag" type. Must contain at least one item.
+      packages:
+      - package-1-name # This is a package name, not an NVR.
+      - package-2-name
+      # Optional. Default value is configured in OSBS environment
+      sigkeys:
+      - AB123456
+
+Alternatively, for modules support::
+
+    compose:
+      type: module
+      # Eventually, ODCS will support ":" separator, for now "-" is used:
+      #     https://pagure.io/odcs/issue/98
+      # Required for "module" type. Must contain at least one item.
+      modules:
+      - "module_name1:stream1"
+      - "module_name2:stream1"
+
+Behavior Changes
+----------------
+
+During an OSBS build, if ``container.yaml`` exists **and** ``compose`` key is defined:
+
+- OSBS orchestrator build will request a compose from ODCS based on specified
+  parameters in ``compose`` section. Once compose completes, build resumes and
+  the URL for the generated compose is passed to worker builds via
+  ``yum_repourls`` build parameter. *NOTE: If odcs-signed parameter was set to
+  false, the compose request will ignore the sigkeys provided in
+  container.yaml.*
+
+- If ``compose`` section sets ``inherit`` to ``true``, the ODCS composes used in
+  building the parent image will be re-used. This may require OSBS to request
+  compose to be regenerated if it’s in *remove* state. Once compose is
+  available, build resumes and the URL for the generated compose is passed to
+  worker builds via ``yum_repourls`` build parameter. *NOTE: this can be used in
+  conjunction with the method stated on previous bullet point. In which case,
+  multiple ODCS composes are used.*
```

Comment:
*How* can multiple ODCS composes by used?

Reply:
This goes together with the `inherit` feature. Potentially, each layered image can generate their own compose as well as re-use the ones from parent.

Reply:
So the idea is that you'd do two composes, inject both yum repourls, and hope for the best? I don't think you achieve "the layered images won't accidentally pick up unrelated newer packages" - any newer packages that are dependencies of the packages listed by the layered image (glibc, etc.) will be pulled into the compose.

[ ODCS seems to always invoke pungi with `koji_inherit = false`, which would help in this regard, but doesn't generally seem reasonable  - probably this just has never been tested in a practical case. ]

Reply:
> [ ODCS seems to always invoke pungi with koji_inherit = false, which would help in this regard, but doesn't generally seem reasonable - probably this just has never been tested in a practical case. ]

This will soon change to actually be true, which I guess further emphasizes your point :)